pkg/cmd/roachtest/tests/sysbench.go

```diff
@@ -281,12 +284,19 @@ type sysbenchMetrics struct {
 	Reconnects   string `json:"reconnects"`
 }
 
+type openmetricsValues struct {
```

Comment:
It feels a bit unfortunate to need this extra struct type and do all the converting from an array to a map for openmetrics. Potential idea:

1. We get rid of `sysbenchMetrics` and just use a `map[string]string`, i.e.:

```
		snapshotTick := map[string]string{
			"time":         strconv.FormatInt(start.Unix(), 10),
			"threads":      fields[1],
			"transactions": fields[3],
			"qps":          fields[5],
			"readQps":      qpsByType[0],
			"writeQps":     qpsByType[1],
			"otherQps":     qpsByType[2],
			"p95Latency":   fields[10],
			"errors":       fields[12],
			"reconnects":   fields[14],
		}
```

I *think* `json.Marshal` would still give the same output for the roachperf case and this makes the openmetrics case a lot easier.

2. Instead of converting all the metrics to bytes at the end, we can mirror what the roachperf logic does and do it per tick. This way we don't need to encode the time for every single metric and can reduce the amount of `t.ExportOpenmetrics()` branching required. i.e.:

```
	  if t.ExportOpenmetrics() {
			snapshotTickBytes = getOpenmetricsBytes(snapshotTick, labelString, start)
		} else {
			snapshotTickBytes, err = json.Marshal(snapshotTick)
			if err != nil {
				return errors.Errorf("error marshaling metrics")
			}
			snapshotTickBytes = append(snapshotTickBytes, []byte("\n")...)
		}
```

```
func getOpenmetricsBytes(
	openmetricsMap map[string]string, labelString string, tick time.Time,
) []byte {
	metricsBuf := bytes.NewBuffer([]byte{})
	for key, value := range openmetricsMap {
		metricsBuf.WriteString(fmt.Sprintf("# TYPE %s gauge\n", key))
		metricsBuf.WriteString(fmt.Sprintf("%s{%s} %s %s\n", key, labelString, value, tick))
	}

	// Add # EOF at the end for openmetrics
	metricsBuf.WriteString("# EOF\n")
	return metricsBuf.Bytes()
}
```

Reply:
Just an idea though, I realize this is touching the existing logic and there might be a desire to minimize impact for that.

Reply:
> Instead of converting all the metrics to bytes at the end, we can mirror what the roachperf logic does and do it per tick. This way we don't need to encode the time for every single metric and can reduce the amount of t.ExportOpenmetrics() branching required. 

I actually thought of doing that in the first time but it was actually polluting the openmetrics file with a lot of `# TYPE` notations at every tick.

Reply:
I have 2 ideas of implementing
1. Create an interface that has tick and export functions for both types that can make the code cleaner.
2. Create separate functions altogether.

I'm not really in favour of option 2 since I might want to reuse the scanner part, and if I write a different functions altogether than there will code repetitions.

wdyt ?

Reply:
Both options are good with me, I suppose thinking about it more it doesn't matter that much if we will eventually remove the non open metrics code. Although with that line of thought, whatever seems easier is probably the way to go.

Reply:
@DarrylWong Yes, the non-openmetrics code will be eventually removed. So should the current changes be good enough or should I go ahead and implement any one of the above two approaches discussed?